### PR TITLE
[efr32] Various modifications to radio driver

### PR DESCRIPTION
**Pull request by dismirlian**
This PR introduces some modifications to the EFR32MG12 radio driver. It is a work in progress, but I would appreciate feedback. I will be pushing more changes during the next days, including "merging" these changes into the MG21 radio driver.

I would also like to propose to unify both the MG12 and MG21 Silabs platforms, since they share a large percentage of code. @nigelg-silabs please see how the nRF52xxx platform drivers are structured.

Thanks.

**Edit:**
**Ready to review**; please see a summary below:

I will not merge these modifications with the MG21 driver for two reasons:
- @nigelg-silabs confirmed there is an ongoing unification effort.
- I don't have MG21 hardware to test.

Modifications/Features:
- Implemented correctly setting `mAckedWithFramePending`.
- Split 2.4GHz and Sub-GHz PA configurations, and introduced means of independently setting power supply to the PA and their ramp times.
- Implemented zero-copy TX.
- Implemented get/set CcaEnergyDetectThreshold.
- Implemented RX timestamping (at the SFD reception).

Bug corrections:
- Corrected `macAckWaitDuration` (from 894us to 864us; it was a typo).
- Honor `mMaxCsmaBackoffs`.
- Removed duplicate calls of RAIL functions which were called once per band, and did not need to (the call was exactly the same for all bands).
- Set `OPENTHREAD_CONFIG_SOFTWARE_CSMA_BACKOFF_ENABLE=0` and   `OPENTHREAD_CONFIG_SOFTWARE_ENERGY_SCAN_ENABLE=0` according to the available hardware capabilities.
- Undefined `RADIO_CONFIG_SRC_MATCH_*_ENTRY_NUM` to let the soft source match table implementation define them as `OPENTHREAD_CONFIG_MLE_MAX_CHILDREN`.
- Signalled event pending when failed to start TX (event pending was not signalled for this case, even if TX operation was actually done).
- Handled promiscuous mode for ACK packets (ACK packets were discarded for sniffer mode).

Minor improvements:
- Use `RAIL_CopyRxPacket` instead of `memcpy` (simpler code).
- Prioritized handling of `RAIL_EVENT_IEEE802154_DATA_REQUEST_COMMAND`.
- Fully removed scheduler if DMP is disabled.
- Simplified handling of debug counters.
- Added performance measurement for source match table lookup.
- Simplified setting channel.
- Used `const` more.
- Used `assert` instead of `otEXPECT` to detect issues when interacting with RAIL.

### CODE REVIEW

**Comment by nigelg-silabs** [comment]
> This PR introduces some modifications to the EFR32MG12 radio driver. It is a work in progress, but I would appreciate feedback. I will be pushing more changes during the next days, including "merging" these changes into the MG21 radio driver.
> 
> I would also like to propose to unify both the MG12 and MG21 Silabs platforms, since they share a large percentage of code. @nigelg-silabs please see how the nRF52xxx platform drivers are structured.
> 
> Thanks.

**Comment by nigelg-silabs** [comment]
We already have work in progress to unify common aspects of the code which predates the nRF52xxx changes.

File: examples/platforms/efr32mg12/brd4166a/board_config.h [hunk 1]
```diff
@@ -36,6 +36,7 @@
 #define __BOARD_CONFIG_H__
 
 #define RADIO_CONFIG_2P4GHZ_OQPSK_SUPPORT 1   /// Dev board suppports OQPSK modulation in 2.4GHz band.
+#define RADIO_CONFIG_2P4GHZ_PA_USES_DCDC  0   /// The 2.4GHz PA is fed from VBAT
```
[the remark this answers is not in the document]
Comment (dismirlian) [reply examples/platforms/efr32mg12/brd4166a/board_config.h]: You are right. I mixed up the schematics of the 4166 and the 4170...

File: examples/platforms/efr32mg12/brd4170a/board_config.h [hunk 2]
```diff
@@ -37,6 +37,8 @@
 
 #define RADIO_CONFIG_2P4GHZ_OQPSK_SUPPORT 1   /// Dev board suppports OQPSK modulation in 2.4GHz band.
 #define RADIO_CONFIG_915MHZ_OQPSK_SUPPORT 1   /// Dev board suppports OQPSK modulation in 915MHz band.
+#define RADIO_CONFIG_2P4GHZ_PA_USES_DCDC  1   /// The 2.4GHz PA is fed from the DCDC
+#define RADIO_CONFIG_SUBGIG_PA_USES_DCDC  1   /// The Sub-GHz PA is fed from the DCDC
```
[the remark this answers is not in the document]
Comment (dismirlian) [reply examples/platforms/efr32mg12/brd4170a/board_config.h]: You are right. I mixed up the schematics of the 4166 and the 4170...

**Comment by dismirlian** [comment]
Hi @andrasbiro, thanks for reviewing!

> * ackTimeout is a good catch!

The new value (864us) is correct in case of beacon-enabled PANs:
macAckWaitDuration = aUnitBackoffPeriod + aTurnaroundTime + phySHRDuration + 6 × phySymbolsPerOctet = 54 symbols = 864us

I think it should be furher reduced to 544us, because OpenThread is based on a beacon-disabled PAN:
macAckWaitDuration = aTurnaroundTime + phySHRDuration + 6 × phySymbolsPerOctet = 34 symbols = 544us

What do you think?


> * the removal of "otherwise only signal MAC layer for non-ACK frame" is probably the right choice, it should be filtered by RAIL_IEEE802154_Config_t.framesMask, but please test it

The following two lines guarantee that `sReceiveFrame.mLength > IEEE802154_ACK_LENGTH` is always true, so the `if` condition can be removed altogether
https://github.com/openthread/openthread/blob/9398c591bfbf7e49bc2010dca23d997975a11ea2/examples/platforms/efr32mg12/radio.c#L711
https://github.com/openthread/openthread/blob/9398c591bfbf7e49bc2010dca23d997975a11ea2/examples/platforms/efr32mg12/radio.c#L753

**Comment by dismirlian** [comment]
See #4358

**Comment by dismirlian** [comment]
Commit 815f8fbfa0b532a5f7fa94d2f294886d582c1285 depends on #4432

**Pull request closed without merging**